delphi/app/data_analyzer.py

```diff
@@ -0,0 +1,244 @@
+        """
+        self.data_keywords          = set() if data_keywords is None else set(data_keywords)
+        self.constraint_keywords    = set() if constraint_keywords is None else set(constraint_keywords)
+
+    def init(self):
+        """
+        Called once before the plugin has recieved any data. Meant to
+        initialize any resources before analyzing.
+        """
+        pass
+
+    def shutdown(self):
+        """
+        Called when the delphi server shutsdown. Meant to clean up
+        any resources that this plugin uses.
+        """
+        pass
+
+    def collect(self):
+        """
+        Called whenever the current state of this data analyzer is needed.
+        
+        Returns:
+            (DataContainer): A map of keywords to values.
+        """
+        pass
+
+    def analyze(self, data):
+        """
+        Called whenever new data has been recieved to analyze. Will only be called
+        if the data payload contains the required keywords that this analyzer is
+        listening for. Will also only get the data payload containing the keywords
+        that this analyzer is listening for.
+
+        Parameters:
+            data (DataContainer): A map of keywords to values.
+        """
+        pass
+
+    def constraint(self, constraints):
+        """
+        Called whenever a new constraint has been recieved. Will only be called
+        if the constraint payload contains the required keywords that this analyzer is
+        listening for.
+
+        Parameters:
+            constraints (ConstraintsContainer): A map of keywords to values.
+        """
+        pass
+
+    def get_data_keywords(self):
+        """
+        Gets all the data keywords this plugin is listening to.
+
+        Returns:
+            (set[string]): keywords
+        """
+        return self.data_keywords
+
+    def get_constraint_keywords(self):
+        """
+        Gets all the constraint keywords this plugin is listening to.
+
+        Returns:
+            (set[string]): keywords
+        """
+        return self.constraint_keywords
+
+    def has_data_keyword(self, keyword):
+        """
+        Check if this plugin is listening to the given keyword in data
+        payloads.
+
+        Parameters:
+            keyword (string): A keyword
+
+        Returns:
+            (bool): True if the keyword is being listened to in data
+                    payloads, false otherwise
+        """
+        return keyword in self.data_keywords
+
+    def has_constraint_keyword(self, keyword):
+        """
+        Check if this plugin is listening to the given keyword in constraint
+        payloads.
+
+        Parameters:
+            keyword (string): A keyword
+
+        Returns:
+            (bool): True if the keyword is being listened to in constraint
+                    payloads, false otherwise
+        """
+        return keyword in self.constraint_keywords
+
+class DataContainer(object):
```

Comment:
The resources we pass around for the data objects and constraint objects should be actual objects not just wrappers around a dictionary. This still has the same issues as a dictionary would have for readability, just with more complexity around it.

Reply:
I am unsure how one would implement a dictionary based problem without using a dictionary, and for that matter how we should be using an object that simulates a dictionary without using a dictionary?